Add new API to allow inserter items to be prioritised

scruffian opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Adds a new API to the block list settings to allow parent blocks to define a priority for blocks in the inserter for a given inner blocks.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

In https://github.com/WordPress/gutenberg/pull/48752 we added a private API to allow blocks to be prioritised. This was so we could force `Page Link` and `Custom Link` to show up as the first two results in the inserter options for the Navigation block's list view editing mode.

However, this was always envisaged as a short term way to allow the wider feature to land in WP 6.2. This PR circles back (and is an alternative to https://github.com/WordPress/gutenberg/pull/49959) with an alternative approach which looks at this problem more holistically by assuming the requirement to prioritise blocks will likely be useful to other block's in future.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds a new `inserterPriority` property to inner blocks which allows the consumer to pass an array of blocks and/or variations which should be prioritised in the Inserter.

It also adds a concrete example, by way of alternating the Nav block to utilise this new property to cause the `Page Link` and `Custom Link` blocks to show up first in the Inserter.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

#### Automated Testing

Run

```
npm run test:e2e packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js
```

#### Manual Testing

- Alter the `INSERTER_PRIORITY` constant in `constants.js` (see changeset) to be:

```js
export const INSERTER_PRIORITY = [
	'core/spacer',
	'core/navigation-link/page',
	'core/navigation-link',
];
```
- Rebuild with `npm run dev`
- New Post
- Add Nav block
- Open inspector controls
- Open List View tab in inspector controls
- Click inserter
- See `core/spacer` block as first item followed by the other two block variations.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

Co-authored-by: Dave Smith <444434+getdave@users.noreply.github.com>

scruffian committed:
Proposing a way to sort items in the block inspector based on allowed blocks

getdave committed:
Add inserterPriority API to inner blocks

getdave committed:
Sort inserter based on inserterPriority prop from block list settings

getdave committed:
Use new inserterPriority API in Nav block

getdave commented:
@scruffian As requested (away from Github) I reworked this to use a new API on inner blocks called `inserterPriority`. 

Why? Because I thought if we reuse `allowedBlocks` then any implementations that currently use `allowedBlocks` will have their inserter items unexpectedly automatically sorted.

Note that the `inserterPriority` items are automatically filtered by the `allowedBlocks` so if it contains blocks/variations that are not supported then they will be removed.

I guess we might need to do the same for the variations though, in case someone includes a variation of a block that isn't in the allowed list.

draganescu commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -108,8 +130,9 @@ function InserterSearchResults( {
 		blockTypes,
 		blockTypeCategories,
 		blockTypeCollections,
-		maxBlockTypes,
```
I see we had `maxBlockTypes` and `maxBlockTypesToShow` both existed in code. Why did we remove it here?

getdave replied:
I think the eslint rule detected a redundant prop. This wasn't due to our change here but was redundant from a previous commit. `maxBlockTypes` is not used within the `useMemo` hook so it's not needed. Perhaps it was used in the past.

draganescu replied:
🆗

draganescu commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -85,14 +89,32 @@ function InserterSearchResults( {
+	const orderInitialBlockItems = useCallback(
+		( items ) => {
+			items.sort( ( { id: aName }, { id: bName } ) => {
+				// Sort block items according to `prioritizedInserterBlocks`.
```
```suggestion
				// Sort block items according to `parentInserterPriority`.
```

getdave committed:
Correct comment

Co-authored-by: Andrei Draganescu <me@andreidraganescu.info>

getdave committed:
Remove redundant prop

ajlende commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -68,9 +70,15 @@ export default function useNestedSettingsUpdate(
 	// render.
 	const _allowedBlocks = useMemo( () => allowedBlocks, allowedBlocks );
 
+	const _inserterPriority = useMemo(
+		() => inserterPriority,
+		inserterPriority
+	);
```
There's no reason for this to be memoized.

And unless `useMemo` handles dependencies differently from `useLayoutEffect`, the previous memo does nothing either. If `allowedBlocks` is getting a new object reference each time, the `_allowedBlocks` will also update to the new reference.

getdave replied:
This is confusing but I think it's correct.

The dependencies list of `useMemo` is an array of dependencies. In this case the dependencies are the strings contained _within_ the `inserterPriority` array, not the `inserterPriority` _array_ itself.

Without the memo, any non-stable `inserterPriority` array would trigger the layout effect because the reference would change.

With the memo, even if the array reference is unstable the memo will fix it to make it stable. This will avoid the layout effect being called unnecessarily.

The memo "works" because the 2nd argument of the `useMemo` call is the _array of strings_. As these strings will stay the same even if the `inserterPriority` array _reference_ changes, the `useMemo` never needs to recompute.

Originally in https://github.com/WordPress/gutenberg/pull/30311/

draganescu replied:
Great explanation! 👏🏻

ajlende replied:
The comment should explain it like that—the existing one isn't very clear when scanning over the code.

getdave replied:
@ajlende Yeh you have a point there. I can raise a PR to augment the existing comments here if you don't fancy taking it on.

getdave commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -426,13 +424,7 @@ export const ComposedPrivateInserter = compose( [
-			orderInitialBlockItems={ undefined }
-		/>
-	);
+	return <ComposedPrivateInserter ref={ ref } { ...props } />;
```
We can probably follow up to remove the need for the private wrapper once (and if!) this PR is merged.

draganescu replied:
I would keep the private wrapper. This shape allows for easy future private APIs without all the boilerplate being re-added.

Mamaduka replied:
I was looking for why we haven't removed the private inserter. IMO, this doesn't seem like a good reason to keep things around just in case we might need them in the future.

getdave commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -70,6 +71,9 @@ function InserterSearchResults( {
 		destinationRootClientId
 	);
 
+	const parentBlockListSettings = getBlockListSettings( rootClientId );
+	const parentInserterPriority = parentBlockListSettings?.inserterPriority;
```
I think we _might_ need to compute this value inside a `useSelect` hook. Why? Because currently if the block list or `rootClientId` were to change then this component would not update. `useSelect` would re-run the selectors if dependencies change so we might need to refactor toward that.

getdave commented:
As this represents a change to a key API, I'm pinging @youknowriad for review as well.

talldan commented on packages/block-editor/src/components/inserter/index.js:
```diff
@@ -426,13 +424,7 @@ export const ComposedPrivateInserter = compose( [
-		<ComposedPrivateInserter
-			ref={ ref }
-			{ ...props }
-			orderInitialBlockItems={ undefined }
```
I suppose the other option is to keep this API, but instead support passing a custom appender to List View (instead of the showAppender prop).

It still doesn't solve the issue where the block itself ends up with different ordering, unless both the block and the list view instance use the same custom appender, so I think what's in this PR is better. It's define once and it works everywhere which means there's less room for errors.

draganescu replied:
I think 

> instead support passing a custom appender to List View (instead of the showAppender prop).

is better than the prop, but why would we need to keep `orderInitialBlockItems`?

getdave replied:
Only if you wanted _specific_ instance X of an inserter to have a specific order.

The new implementation in this PR is a blanket application. Any block that uses the new prop on `InnerBlocks` no matter where it is with have items prioritised.

The previous implementation was more targetted.

draganescu replied:
Wouldn't `orderInitialBlockItems` be the default order for when block's don't prioritize or for the generic canvas inserter in an empty paragraph?

Edit: no, that's for `orderInitialBlockItems` in the search results ro handle. This `orderInitialBlockItems` prop can be removed because it's effectively replaced by a public API doing the same thing.

talldan reviewed (commented):
Thank you for exploring this. It may have taken longer, but it feels like the better option to me.

I didn't have a chance to do a full code review, but the code looks like what I'd expect.

I'm not sure if it could be made private, but it also feels like a really useful feature to me, so it could shipped stable. Tests and docs might be needed once there's an agreement on this.

youknowriad commented:
If I'm not wrong @ntsekouras did some work in the past for internal ordering in the inserter. So he might have some feedback here.

getdave committed:
Avoid stale inserterPriority

getdave committed:
Make sorting function stable

draganescu commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -31,6 +33,25 @@ const INITIAL_INSERTER_RESULTS = 9;
+		return items;
+	}
+
+	// Sort is "in place".
```
What does this comment mean?

getdave replied:
Haha. Sorry yes. I should have removed that. It was to remind me that `.sort()` will _mutate_ rather than return a new reference like things like `.filter()` would do.

draganescu commented:
This looks excellent IMO. it solves the block inserter being tailored to what the block accepts as inner blocks and to what the most common UX is.

ntsekouras commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -46,10 +67,21 @@ function InserterSearchResults( {
 } ) {
 	const debouncedSpeak = useDebounce( speak, 500 );
 
+	const { parentInserterPriority } = useSelect(
```
Nits: No need to return an object here and also no need to prepend the `parent` in the var name IMO.

getdave replied:
Counter argument is expressed in this post under heading `Prefer returning objects with properties from the callback`

https://jsnajdr.wordpress.com/2021/01/22/some-best-practices-for-using-useselect-from-wordpress-data/#:~:text=%7D-,Prefer,-returning%20objects%20with

ntsekouras replied:
No strong opinion here :) . In our case the `prioritizedInserterBlocks` though is a `const` and doesn't change.

ntsekouras commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -89,10 +121,16 @@ function InserterSearchResults( {
-		if ( ! filterValue && orderInitialBlockItems ) {
-			orderedItems = orderInitialBlockItems( orderedItems );
+
+		if ( ! filterValue && parentInserterPriority ) {
```
This functionality would be great to add in slash inserter too [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/autocompleters/block.js#L67).

getdave replied:
I'll take a look at this.

getdave replied:
This would involve extracting a hook for the retrieval of `prioritisedInserterBlocks` and also moving the sorting function to a shared utils location.

Once done this should be simple enough. I can tackle tomorrow.

getdave replied:
This would involve extracting a hook for the retrieval of `prioritisedInserterBlocks` and also moving the sorting function to a shared utils location.

Once done this should be simple enough. I can tackle tomorrow.

scruffian replied:
> This functionality would be great to add in slash inserter too [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/autocompleters/block.js#L67).

Let's do this in a followup.

ntsekouras replied:
We can, but in general this is a tiny change. I tried it before commenting. We just need to find a good place for the sorting function(ex utils or something). Also we don't even need a new hook, as is just a `select` call.

scruffian replied:
Created an issue here: https://github.com/WordPress/gutenberg/issues/50635

ntsekouras commented on packages/block-library/src/navigation/edit/inner-blocks.js:
```diff
@@ -93,6 +93,7 @@ export default function NavigationInnerBlocks( {
 			onInput,
 			onChange,
 			allowedBlocks: ALLOWED_BLOCKS,
+			inserterPriority: INSERTER_PRIORITY,
```
I think a different name more close to what it actually is, would be better. Maybe `prioritizedInserterBlocks` or something like that..

getdave replied:
Ok no problem.

getdave replied:
✅

ntsekouras reviewed (commented):
Thanks for the PR @scruffian! I haven't tested thoroughly, but it looks good.

getdave committed:
Renaming

getdave committed:
Remove redundant comment

MaggieCabrera reviewed (approved):
This is testing nicely, the e2e tests were passing locally. Let's see if GH can handle those today...

MaggieCabrera commented on packages/block-library/src/navigation/constants.js:
```diff
@@ -14,3 +14,9 @@ export const ALLOWED_BLOCKS = [
 ];
+
+export const INSERTER_PRIORITY = [
+	'core/spacer',
```
did we mean to put the spacer here or is this for testing?

getdave replied:
Nice catch 😓

scruffian committed:
remove spacer

ajlende commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -66,11 +68,22 @@ export default function useNestedSettingsUpdate(
 
 	// Memoize as inner blocks implementors often pass a new array on every
 	// render.
```
```suggestion
	// Memoize allowedBlocks and prioritisedInnerBlocks based on the contents
	// of the arrays. Implementors often pass a new array on every render,
	// and the contents of the arrays are just strings, so the entire array
	// can be passed as dependencies.
	
```

ntsekouras commented:
Please before merging this, docs and some tests should be added, as is a new API.

getdave commented:
A good next step here would be to identify whether there are any tests covering the `allowedBlocks` feature? If not then we will need to craft tests for this new API from scratch.

jeryj committed:
Set prioritisedBlocks as empty array when no blockListSettings are found.

There are instances in the main inserter search results where the rootClientId is undefined, such as when searching for a block. This means there are no blockListSettings, which ends up in no `proritisedBlocks` property being returned and it crashses the app.

jeryj committed:
proritise -> prioritize for consistency

scruffian committed:
Update packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js

Co-authored-by: Alex Lende <alex@lende.xyz>

jeryj committed:
Renaming constant to match updated name

jeryj committed:
Add prioritizedKInnerBlocks to the inner-blocks README

scruffian committed:
lint fix

scruffian committed:
update comment

Mamaduka commented on packages/block-editor/src/components/inserter/search-results.js:
```diff
@@ -46,10 +66,22 @@ function InserterSearchResults( {
+
+			return {
+				prioritizedBlocks:
+					blockListSettings?.prioritizedInserterBlocks || [],
```
```suggestion
					blockListSettings?.prioritizedInserterBlocks || EMPTY_ARRAY,
```

It's preferred to return stable references.

Mamaduka commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -64,13 +66,27 @@ export default function useNestedSettingsUpdate(
-	// Memoize as inner blocks implementors often pass a new array on every
-	// render.
-	const _allowedBlocks = useMemo( () => allowedBlocks, allowedBlocks );
+	// Memoize allowedBlocks and prioritisedInnerBlocks based on the contents
+	// of the arrays. Implementors often pass a new array on every render,
+	// and the contents of the arrays are just strings, so the entire array
+	// can be passed as dependencies.
+
+	const _allowedBlocks = useMemo(
+		() => allowedBlocks,
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		allowedBlocks
+	);
+
+	const _prioritizedInserterBlocks = useMemo(
+		() => prioritizedInserterBlocks,
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		prioritizedInserterBlocks
+	);
```
Note: We'll have to remove this hack in the futere if we want to support dynamic `allowedBlocks` and `prioritizedInserterBlocks`. See https://github.com/WordPress/gutenberg/issues/14515#issuecomment-1429218762.

scruffian committed:
update comment

MaggieCabrera commented:
> A good next step here would be to identify whether there are any tests covering the `allowedBlocks` feature? If not then we will need to craft tests for this new API from scratch.

not the `allowedBlocks` specifically, but the correct order of blocks in the inserter is covered by some navigation tests already [here](https://github.com/WordPress/gutenberg/blob/f4da24145ec853cbb94f8f5800f37c13f725bd0e/test/e2e/specs/editor/blocks/navigation.spec.js#L539)

scruffian committed:
pass the correct props to useNestedSettingsUpdate

getdave committed:
Use stable ref

Co-authored-by: George Mamadashvili <georgemamadashvili@gmail.com>

getdave commented on packages/block-editor/src/components/inner-blocks/index.native.js:
```diff
@@ -73,9 +73,13 @@ function UncontrolledInnerBlocks( props ) {
 		clientId,
 		allowedBlocks,
+		prioritizedInserterBlocks,
+		__experimentalDefaultBlock,
+		__experimentalDirectInsert,
 		template,
 		templateLock,
 		templateInsertUpdatesSelection,
+		__experimentalCaptureToolbars: captureToolbars,
```
Do we _need_ these ones?

draganescu replied:
I don't understand this question.

getdave replied:
The question is "why are we adding all these extra props when this PR is about `prioritizedInserterBlocks`"?

getdave commented:
Just to flag up the follow items that are "blocking" the merge of this PR:

- native is broken and we'll need to fix that
- we need to check whether current test coverage specifically tests the new changes in this PR (doubtful)
- consider whether to [add the same functionality to the slash inserter](https://github.com/WordPress/gutenberg/pull/50510#discussion_r1191518572)
- write docs

fluiddot commented:
👋 I added the label `Mobile App - Automation` to create [this Gutenberg Mobile PR automatically](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770). The Gutenberg Mobile repo is where we run most of the E2E tests for the mobile native version, so it will tell us if these changes might have affected the app. We'll share here if all tests pass.

MaggieCabrera reviewed (dismissed):
I thought I could dismiss my review but I can't, so let's hold off on merging until we address @getdave [comments](https://github.com/WordPress/gutenberg/pull/50510#issuecomment-1545453335)

MaggieCabrera commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -180,3 +180,8 @@ For example, a button block, deeply nested in several levels of block `X` that u
 -   **Type:** `Function`
 -   **Default:** - `undefined`. The placeholder is an optional function that can be passed in to be a rendered component placed in front of the appender. This can be used to represent an example state prior to any blocks being placed. See the Social Links for an implementation example.
+
+### `prioritizedInserterBlocks`
```
do we need to document this somewhere else or is this enough?

getdave replied:
That's the official docs so I guess that's enough.

getdave committed:
Register the test Plugin for e2e tests

getdave committed:
Register block with prioritzedInserterBlocks set

getdave committed:
Add initial test scaffold

getdave committed:
Tidy up scaffolded test

getdave committed:
Add test for new API

It fails :(

getdave committed:
Try removing sort from helper

Why is this even here? It shouldn’t transfer results like this.

getdave committed:
Fix test

getdave commented on packages/e2e-test-utils/src/get-all-block-inserter-item-titles.js:
```diff
@@ -20,5 +20,5 @@ export async function getAllBlockInserterItemTitles() {
 		} );
 	} );
-	return [ ...new Set( inserterItemTitles ) ].sort();
+	return [ ...new Set( inserterItemTitles ) ];
```
I changed this in https://github.com/WordPress/gutenberg/pull/50510/commits/8cb75c86b57bb06360fc279cf21897c18053bd53.

Because it does sorting on the results the tests cannot assert on the priority of the blocks.

This helper is "doing it wrong". Instead it should return raw results and the _consumer_ should do the sorting.

Let's see which tests break before deciding how to proceed from here.

getdave replied:
Do we think this should be it's own sub PR?

getdave committed:
Add test to check does not override allowedBlocks when conflicted

getdave committed:
Add additional assertion for retaining of correct number of results

getdave committed:
Ensure tests reflect target of Quick Inserter

MaggieCabrera committed:
sort allowed blocks on the tests that consume getAllBlockInserterItemTitles

fluiddot commented:
> 👋 I added the label `Mobile App - Automation` to create [this Gutenberg Mobile PR automatically](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770). The Gutenberg Mobile repo is where we run most of the E2E tests for the mobile native version, so it will tell us if these changes might have affected the app. We'll share here if all tests pass.

All E2E tests passed in the [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5770) ([CI job reference](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/20287/workflows/27266789-e147-4e5a-9a77-5b1405b997ca)) 🎊 . Note that the PR gets automatically updated when a new commit is pushed in this PR, so the tests will be re-triggered with next commits.

draganescu commented on packages/block-editor/src/components/inner-blocks/index.native.js:
```diff
@@ -97,7 +101,17 @@ function UncontrolledInnerBlocks( props ) {
+		templateLock,
+		captureToolbars,
+		orientation,
+		layout
```
Why is this PR adding all these other settings and not only the core feature `prioritizedInserterBlocks`?

getdave replied:
Because not doing that broke the Mobile implementation (cc @fluiddot).

draganescu replied:
Ah cool!

draganescu commented on packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js:
```diff
@@ -0,0 +1,81 @@
+		style: { outline: '1px solid gray', padding: 5 },
+	};
+
+    // Make it easier to select the block.
```
What does this comment mean?

getdave replied:
It was indicating that without a placeholder within the innerblocks it can be difficult to select the block using e2e tests, especially using Puppeteer. Open to suggestions for improving the comment 🙇

draganescu replied:
> without a placeholder within the innerblocks it can be difficult to select the block using e2e tests, especially using Puppeteer, so we use an image block which has a placeholder.

draganescu reviewed (approved):
I think this is in great shape. Left a few comments but nothing blocking 👏🏻

getdave commented on packages/e2e-tests/plugins/inner-blocks-prioritized-inserter-blocks/index.js:
```diff
@@ -0,0 +1,81 @@
+		style: { outline: '1px solid gray', padding: 5 },
+	};
+
+    // Make it easier to select the block.
```
```suggestion
    // without a placeholder within the inner blocks it can be difficult to select the block using e2e tests
    // especially using Puppeteer, so we use an image block which has a placeholder.
```

getdave committed:
Improve e2e test comment

ntsekouras commented on packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js:
```diff
@@ -25,6 +25,7 @@ const pendingSettingsUpdates = new WeakMap();
  * @param {string}               clientId                   The client ID of the block to update.
  * @param {string[]}             allowedBlocks              An array of block names which are permitted
  *                                                          in inner blocks.
+ * @param {string[]}             prioritizedInserterBlocks  Block names and/or block variations to be prioritized in the inserter.
```
We should be a bit more specific how a block variation is expressed in this API, and this applies to the docs above. We use a convention of `{blockName}/{variationName}` due to the inserter [items structure](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/store/selectors.js#L1816)

scruffian replied:
```suggestion
 * @param {string[]}             prioritizedInserterBlocks  Block names and/or block variations to be prioritized in the inserter, in the format {blockName}/{variationName}.
```

scruffian committed:
Update packages/block-editor/src/components/inner-blocks/use-nested-settings-update.js

scruffian commented on packages/block-editor/src/components/inner-blocks/README.md:
```diff
@@ -180,3 +180,8 @@ For example, a button block, deeply nested in several levels of block `X` that u
+### `prioritizedInserterBlocks`
+
+-   **Type:** `Array`
+-   **Default:** - `undefined`. Determines which inner blocks should be returned first from the block inserter. For example, when inserting a block within the Navigation Block, `core/navigation-link` and `core/navigation-link/page` are the most commonly used inner blocks. We can use `prioritizedInserterBlocks` to pass these `navigation-link` blocks as an array so they can be returned first by def ...
```
```suggestion
-   **Default:** - `undefined`. Determines which block types should be shown in the block inserter. For example, when inserting a block within the Navigation block we specify `core/navigation-link` and `core/navigation-link/page` as these are the most commonly used inner blocks. `prioritizedInserterBlocks` takes an array of the form {blockName}/{variationName}, where {variationName} is optional.
```

scruffian committed:
Update packages/block-editor/src/components/inner-blocks/README.md

scruffian commented:
I updated the docs. I say lets merge this and make sure to follow up with the slash inserter fix ASAP.

ntsekouras commented on packages/e2e-tests/specs/editor/plugins/inner-blocks-prioritized-inserter-blocks.test.js:
```diff
@@ -0,0 +1,108 @@
+
+			await page.waitForSelector( QUICK_INSERTER_RESULTS_SELECTOR );
+
+			await expect( await getAllBlockInserterItemTitles() ).toHaveLength(
```
I noticed this test in the follow up PR and it doesn't seem to add any value. We should either test something about the existence/non-existence of some blocks or remove it.

getdave replied:
Yes I agree. Seems pointless. It was probably me copying boilerplate code about.

We should probably change it to test what happens if you pass an empty value or empty array as the prioritized blocks argument.
